### PR TITLE
Add ChaCha20 encryption option (XOR)

### PR DESCRIPTION
**Pull request by jonasschnelli**
The current ChaCha20 implementation does not support message encryption (it can only output the keystream which is sufficient for the RNG).

This PR adds the actual XORing of the `plaintext` with the `keystream` in order to return the desired `ciphertext`.

Required for v2 message transport protocol.

### CODE REVIEW

File: src/crypto/chacha20.cpp [hunk 1]
```diff
@@ -100,6 +102,10 @@ void ChaCha20::Output(unsigned char* c, size_t bytes)
 
     for (;;) {
         if (bytes < 64) {
+            if (m != nullptr) {
+                for (i = 0;i < bytes;++i) tmp[i] = m[i];
```
Comment (ryanofsky) [inline src/crypto/chacha20.cpp]: A comment would be helpful here: "If m has fewer than 64 bytes available, copy m to tmp and read from tmp instead"

Comment (ryanofsky) [inline src/crypto/chacha20.cpp]: Just `memcpy(tmp, m, bytes)` might be simpler and easy to read than this for loop.

File: src/crypto/chacha20.cpp [hunk 2]
```diff
@@ -146,6 +152,25 @@ void ChaCha20::Output(unsigned char* c, size_t bytes)
         x14 += j14;
         x15 += j15;
 
+        if (m != nullptr) {
+            x0 = XOR(x0, ReadLE32(m + 0));
```
Comment (ryanofsky) [inline src/crypto/chacha20.cpp]: Not sure what the benefit of defining XOR macro is. Isn't `x0 ^= ReadLE32(m + 0)` more readable and less error-prone than `x0 = XOR(x0, ReadLE32(m + 0))`?

File: src/crypto/chacha20.h [hunk 3]
```diff
@@ -20,7 +20,9 @@ class ChaCha20
     void SetKey(const unsigned char* key, size_t keylen);
     void SetIV(uint64_t iv);
     void Seek(uint64_t pos);
-    void Output(unsigned char* output, size_t bytes);
+
+    // if <input> is provided, the ciphertext will be written to <output> otherwise the current keystream
```
Comment (ryanofsky) [inline src/crypto/chacha20.h]: This comment seems imply that output is not written if input is null. Maybe rephrase: "If \<input> is non-null, the specified number of \<bytes> from \<input> will encrypted and written to \<output>, otherwise just the keystream will be written to \<output\> (which may be useful for random number generation). Either way, the number of bytes written to \<output> will be \<bytes>."

File: src/test/crypto_tests.cpp [hunk 4]
```diff
@@ -187,16 +187,19 @@ static void TestAES256CBC(const std::string &hexkey, const std::string &hexiv, b
...
     std::vector<unsigned char> key = ParseHex(hexkey);
+    std::vector<unsigned char> m = ParseHex(hex_message);
     ChaCha20 rng(key.data(), key.size());
     rng.SetIV(nonce);
     rng.Seek(seek);
     std::vector<unsigned char> out = ParseHex(hexout);
     std::vector<unsigned char> outres;
     outres.resize(out.size());
-    rng.Output(outres.data(), outres.size());
+
```
Comment (ryanofsky) [inline src/test/crypto_tests.cpp]: Maybe `assert(hex_message.empty() || m.size() == out.size())`

File: src/test/crypto_tests.cpp [hunk 5]
```diff
@@ -187,16 +187,19 @@ static void TestAES256CBC(const std::string &hexkey, const std::string &hexiv, b
...
+    // perform the ChaCha20 round(s), if message is provided it will output the encrypted ciphertext otherwise the keystream
+    rng.Output(hex_message.empty() ? nullptr : m.data(), outres.data(), outres.size());
```
Comment (ryanofsky) [inline src/test/crypto_tests.cpp]: It seems like a useful additional test would be call `rng.Output` twice if `hex_message` is nonempty, once with input and once without, and confirm that xoring the two outputs yields the input again.

**Review by ryanofsky — APPROVED** [review]
utACK 4325da919a042dfb4eed87b6d038dbb236959e5b. I don't know very much about chacha20 or this method of encryption, but this seems to do what is described.

File: src/crypto/chacha20.cpp (hunk 1 above)
Comment (jonasschnelli) [reply src/crypto/chacha20.cpp]: I'm not to familiar with compiler optimisations and stuff but I just checked the reference implementation by DJB and some other C based ChaCha implementations and all of them seems to use the byte per byte assignment. I prefer to leave it as it is.

File: src/test/crypto_tests.cpp (hunk 5 above)
Comment (jonasschnelli) [reply src/test/crypto_tests.cpp]: Good point. Added.

File: src/test/crypto_tests.cpp (hunk 4 above)
Comment (jonasschnelli) [reply src/test/crypto_tests.cpp]: Added.

File: src/crypto/chacha20.h (hunk 3 above)
Comment (jonasschnelli) [reply src/crypto/chacha20.h]: Valid point. Fixed.

File: src/crypto/chacha20.cpp (hunk 1 above)
Comment (jonasschnelli) [reply src/crypto/chacha20.cpp]: Added.

File: src/crypto/chacha20.cpp (hunk 2 above)
Comment (jonasschnelli) [reply src/crypto/chacha20.cpp]: Thanks for the review! Your right. Changed.

File: src/crypto/chacha20.cpp [hunk 6]
```diff
@@ -71,7 +71,7 @@ void ChaCha20::Seek(uint64_t pos)
     input[13] = pos >> 32;
 }
 
-void ChaCha20::Output(unsigned char* c, size_t bytes)
+void ChaCha20::Output(const unsigned char* m, unsigned char* c, size_t bytes)
```
Comment (sipa) [inline src/crypto/chacha20.cpp]: I think it would be better to duplicate the logic here and turn the stream cipher version into a separate function.

As is, this will do double work in outputting both, and introduce an unnecessary branch in the output version.

Comment (jonasschnelli) [reply src/crypto/chacha20.cpp]: Good point. Adapted the duplication approach.

File: src/crypto/chacha20.cpp [hunk 7]
```diff
@@ -178,3 +178,135 @@ void ChaCha20::Output(unsigned char* c, size_t bytes)
...
+        x8 += j8;
+        x9 += j9;
+        x10 += j10;
+        x11 += j11;
+        x12 += j12;
+        x13 += j13;
+        x14 += j14;
+        x15 += j15;
+
+        if (m != nullptr) {
```
Comment (sipa) [inline src/crypto/chacha20.cpp]: No need for this `if`.

Comment (jonasschnelli) [reply src/crypto/chacha20.cpp]: Yes. Fixed.

**Review by ryanofsky — APPROVED** [review]
utACK 427b49dcdbfb25153ee4228dda49aa07db16d59c. Changes since last review: splitting Output/Crypt functions, dropping comment describing function arguments, adding comment describing m/tmp copy, dropping XOR macro, and making suggested test changes.

File: src/crypto/chacha20.cpp (hunk 6 above)
Comment (jnewbery) [reply src/crypto/chacha20.cpp]: What's the reasoning here? Performance?

How does the version in the refrence implementation compare?

```c
void ECRYPT_keystream_bytes(ECRYPT_ctx *x,u8 *stream,u32 bytes)
{
  u32 i;
  for (i = 0;i < bytes;++i) stream[i] = 0;
  ECRYPT_encrypt_bytes(x,stream,stream,bytes);
}
```

ie setting `Output(c, bytes)` to call `Crypt(c, c, bytes)`?

File: src/crypto/chacha20.h [hunk 8]
```diff
@@ -20,7 +20,9 @@ class ChaCha20
     void SetKey(const unsigned char* key, size_t keylen);
     void SetIV(uint64_t iv);
     void Seek(uint64_t pos);
-    void Output(unsigned char* output, size_t bytes);
+
+    void Output(unsigned char* c, size_t bytes);
+    void Crypt(const unsigned char* input, unsigned char* output, size_t bytes);
```
Comment (jnewbery) [inline src/crypto/chacha20.h]: Now that this class can encrypt plaintext, the class comment should be updated from "A PRNG class for ChaCha20."

Really, I'd like to see more commenting for this class in general (ie note that `SetIV` sets the nonce, `Seek` sets the counter, that the counter is 64 bytes and the nonce is 64 bytes as in the original ChaCha20 spec instead of 32 byte counter/96 byte nonce as in the IETF spec).

(there are a couple of good resources on the different counter/nonce sizes here: https://libsodium.gitbook.io/doc/advanced/stream_ciphers/chacha20 and here: http://loup-vaillant.fr/tutorials/chacha20-design)

File: src/crypto/chacha20.h [hunk 9]
```diff
@@ -20,7 +20,9 @@ class ChaCha20
     void SetKey(const unsigned char* key, size_t keylen);
     void SetIV(uint64_t iv);
     void Seek(uint64_t pos);
-    void Output(unsigned char* output, size_t bytes);
+
+    void Output(unsigned char* c, size_t bytes);
```
Comment (jnewbery) [inline src/crypto/chacha20.h]: nit: This function should be renamed `Keystream` now that the class is not just used for PRNG

File: src/test/crypto_tests.cpp [hunk 10]
```diff
@@ -187,17 +187,36 @@ static void TestAES256CBC(const std::string &hexkey, const std::string &hexiv, b
...
+    assert(hex_message.empty() || m.size() == out.size());
+
+    // perform the ChaCha20 round(s), if message is provided it will output the encrypted ciphertext otherwise the keystream
+    if (!hex_message.empty()) {
+        rng.Crypt(m.data(), outres.data(), outres.size());
+    } else {
+        rng.Output(outres.data(), outres.size());
+    }
     BOOST_CHECK(out == outres);
+    if (!hex_message.empty()) {
```
Comment (jnewbery) [inline src/test/crypto_tests.cpp]: This test seems redundant to me. How could it ever fail if the test above succeeds?

**Review by jnewbery — COMMENTED** [review]
Implementation and tests look good and match the reference implementation here: https://cr.yp.to/streamciphers/timings/estreambench/submissions/salsa20/chacha8/merged/chacha.c and test vector here: https://tools.ietf.org/html/rfc7539#section-2.4.2

I have a few minor comments inline, and would be curious about the answer to:
https://github.com/bitcoin/bitcoin/pull/15512#discussion_r280833386

File: src/test/crypto_tests.cpp (hunk 10 above)
Comment (ryanofsky) [reply src/test/crypto_tests.cpp]: re: https://github.com/bitcoin/bitcoin/pull/15512#discussion_r280836767

Not sure exactly what you want to get rid of, but if you're talking about the `out == outres` check below, I suggested it and think it'd be better to keep, because it means if the test passes you can be reassured that Crypt() and Output() results are consistent without manually verifying `4c616 ^ 6e2e3 == 224f5`... in the test cases.

File: src/crypto/chacha20.h (hunk 9 above)
Comment (jonasschnelli) [reply src/crypto/chacha20.h]: Renamed to `Keystream`

File: src/crypto/chacha20.h (hunk 8 above)
Comment (jonasschnelli) [reply src/crypto/chacha20.h]: The IETF has a slightly modified recommendation AFAIK.
We use Bernstein-Vanilly 64bit nonce and 64bit block counter.

* Changed the class comment
* Added some function comments

File: src/bench/chacha20.cpp [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+static void CHACHA20(benchmark::State& state, size_t buffersize)
+{
+    std::vector<uint8_t> key(32,0);
+    ChaCha20 ctx(key.data(), key.size());
+    ctx.SetIV(0);
+    ctx.Seek(0);
+    std::vector<uint8_t> in(buffersize,0);
+    std::vector<uint8_t> out(buffersize,0);
+    while (state.KeepRunning())
+        ctx.Crypt(in.data(), out.data(), in.size());
```
Comment (jnewbery) [inline src/bench/chacha20.cpp]: micronit: same line or braces for while code block please! Same for while block in `HASH()` below.

File: src/bench/chacha20.cpp [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+static void HASH(benchmark::State& state, size_t buffersize)
```
Comment (jnewbery) [inline src/bench/chacha20.cpp]: Is a reason you've added these `CHash256` benchmarks to chacha20.cpp?

**Review by jnewbery — COMMENTED** [review]
Looks great. Thanks for adding the comments to chacha20.h!

Just a couple of nits in the bench file. I'm still curious about the code duplication between `Keystream()` and `Crypt()`.

File: src/bench/chacha20.cpp (hunk 12 above)
Comment (jonasschnelli) [reply src/bench/chacha20.cpp]: I added if for comparison and the impact on the networking... though I think it belong to an extra commit in https://github.com/bitcoin/bitcoin/pull/15649
Will remove it from here.

File: src/crypto/chacha20.cpp (hunk 6 above)
Comment (jonasschnelli) [reply src/crypto/chacha20.cpp]: I had it in the initial version like @jnewbery just proposed.
After @sipa recommended to separate it, I thought that a strict separation may be beneficial for verification and later optimizations. But maybe @sipa can clarify...

File: src/bench/chacha20.cpp (hunk 11 above)
Comment (jonasschnelli) [reply src/bench/chacha20.cpp]: Fixed

**Comment by jnewbery** [comment]
Looks good. utACK 2dfe2751713c814aea53b5a7563eb74ad1baea00.

In general, it's better to have less code duplication, so I'd like to hear from @sipa his reasoning for https://github.com/bitcoin/bitcoin/pull/15512#discussion_r269692281

**Comment by sipa** [comment]
@jnewbery I just thought that it'd be preferable not to burden the RNG code with branches that are only used for encryption. In cryptographic code like this, I don't care too much about duplication, as it isn't code that subject to many possible future changes.

**Comment by jnewbery** [comment]
ok, my aesthetic preference is for less duplication, either by having a single Keystream/Crypt function, or by calling Keystream() from Crypt() with a zero'ed message. As sipa points out though, this code is unlikely to be modified much in future, so it's not a big deal either way.

utACK 2dfe2751713c814aea53b5a7563eb74ad1baea00

**Review by ryanofsky — APPROVED** [review]
utACK 2dfe2751713c814aea53b5a7563eb74ad1baea00. Changes since last review are just renaming the Crypt method, adding comments, and simplifying the benchmark.

> I just thought that it'd be preferable not to burden the RNG code with branches that are only used for encryption.

I think the duplication is fine, but it's unclear if the concern with branches is about readability or performance. If it's performance, you could give the previous `Output()` function an additional `template<bool use_input>` template argument, and have `Crypt()` and `Keystream()` both call it inlined.

**Comment by sipa** [comment]
utACK 2dfe2751713c814aea53b5a7563eb74ad1baea00

File: src/crypto/chacha20.cpp [hunk 13]
```diff
@@ -178,3 +178,133 @@ void ChaCha20::Output(unsigned char* c, size_t bytes)
...
+        WriteLE32(c + 40, x10);
+        WriteLE32(c + 44, x11);
+        WriteLE32(c + 48, x12);
+        WriteLE32(c + 52, x13);
+        WriteLE32(c + 56, x14);
+        WriteLE32(c + 60, x15);
+
+        if (bytes <= 64) {
+            if (bytes < 64) {
+                for (i = 0;i < bytes;++i) ctarget[i] = c[i];
```
Comment (Kixunil) [inline src/crypto/chacha20.cpp]: Isn't this no-op? When `bytes < 64` line 215 get's executed which makes ctarget and c equal pointers, therefore it's copying a value to itself.

File: src/crypto/chacha20.h [hunk 14]
```diff
@@ -17,10 +18,17 @@ class ChaCha20
 public:
     ChaCha20();
     ChaCha20(const unsigned char* key, size_t keylen);
-    void SetKey(const unsigned char* key, size_t keylen);
-    void SetIV(uint64_t iv);
-    void Seek(uint64_t pos);
-    void Output(unsigned char* output, size_t bytes);
+    void SetKey(const unsigned char* key, size_t keylen); //!< set key with flexible keylength; 256bit recommended */
```
Comment (Kixunil) [inline src/crypto/chacha20.h]: I like the renaming. Would be nice if the comments included information about pointers (e.g. "key must be non-null, pointer isn't stored").

File: src/crypto/chacha20.cpp (hunk 13 above)
Comment (jonasschnelli) [reply src/crypto/chacha20.cpp]: I don't think so.
If we do less then 64 bytes, we execute the round on `tmp` (because we do 64bytes always, see line 216, `c` points to `tmp` if <64 bytes). So in the case of <64 bytes, `c` points to the temporary 64byte buffer and `ctarget` points to the function provided `c` argument.

File: src/crypto/chacha20.h (hunk 14 above)
Comment (jonasschnelli) [reply src/crypto/chacha20.h]: Yes. That would have been useful. Feel free to PR.

File: src/crypto/chacha20.cpp (hunk 13 above)
Comment (Kixunil) [reply src/crypto/chacha20.cpp]: Ah, yes, missed that. Thanks and sorry for bothering.
